WordPress/gutenberg
# [RNMobile] Add Contrast Checker to Text-Based Blocks

Fixes https://github.com/WordPress/gutenberg/issues/33396

`Gutenberg Mobile:` https://github.com/wordpress-mobile/gutenberg-mobile/pull/4357

## Description

With this PR, a contrast checker is added to the text-based blocks in the native apps. This checker will display the following notice when users pick a custom text/background colours with poor contrast:

```
This colour combination may be hard for people to read. Try using a brighter background colour and/or a darker text colour.
```

## How has this been tested?

* Add a new text block (e.g. the paragraph block) to a new a post in the mobile app and type some text.
* Tap the block's cog/gear icon to edit its settings.
* Select a foreground and background colour that have a low contrast to one another (e.g. two really dark colours).
* The new notice should appear to inform you that the contrast is low between the two selected colours.

## Testing Notes

* If a theme _isn't_ block-based then the contrast checker will only run when both a custom text and a custom background colour are set for a block i.e. it won't run if a default colour is being used. This is because there isn't any ideal or straightforward way to determine the default colours of a theme that _isn't_ block-based. The web is able to use [getComputedStyles()](https://developer.mozilla.org/en-US/docs/Web/API/Window/getComputedStyle) but this isn't an option for native development. 
* As this change is intended to help accessibility, we should ensure that the notice itself is accessible. The contrast should meet AA standards and be readable via screen readers.
* To keep the scope of the PR narrow, the notice is only introduced to the main settings screen of blocks that have a colour panel in their settings. The notice could be added to individual screens in future iterations.

## Screenshots 

| Light mode  | Dark mode |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/145132004-c8f7a3ea-128d-4207-bf28-74f239f6a801.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/145125553-321dba53-533d-4847-9e06-e9b5773996c5.png" width="100%"> |

## Video

https://user-images.githubusercontent.com/2998162/145385504-bc1a1872-1d7b-4959-a290-abc9cbda6d38.mov

## Types of change

New feature (non-breaking change which adds functionality), with the following high-level overview of the code:

* A native version of [`components/contrast-checker`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/contrast-checker/README.md) has been created for use with the apps.
* In addition, a native version of [`hooks/color-panel`](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-editor/src/components/colors) has been created. This was necessary in order to detect active text/background colours from the apps.
*  `components/contrast-checker` is called from `hooks/color-panel.native` in order to run the checker within blocks that call that hook.






## Checklist:
